### PR TITLE
IIIF Support

### PR DESCRIPTION
**Pull request by edsilv**
Starting a PR to include libvips' new IIIF support as per: https://github.com/libvips/libvips/issues/1465#issuecomment-552487666

However, with the changes included in this PR as per https://github.com/lovell/sharp/issues/1335#issuecomment-552613752 it doesn't generate the tiles and info.json...

I thought passing the correct layout should now work? e.g.

```
    sharp("test.jpg")
    .jpeg()
    .tile({
      size: 512,
      layout: "iiif"
    })
    .toFile("info.json");
```

As far as I understand it, the layout should be getting passed here? 

https://github.com/lovell/sharp/blob/master/src/pipeline.cc#L966

### CODE REVIEW

**Comment by lovell** [comment]
Hi Edward, thank you for the PR, I think we'll need to change this logic too:

https://github.com/lovell/sharp/blob/d5ecc537afa9b5329e79d4a6e50b945dfcf2fcc6/src/pipeline.cc#L1380-L1387

Rather than if-else-if-else... we should probably change this whole block to something simpler like the following (untested):

```cc
baton->tileLayout = static_cast<VipsForeignDzLayout>(
    vips_enum_from_nick(nullptr, VIPS_TYPE_FOREIGN_DZ_LAYOUT,
    sharp::AttrAsStr(options, "tileLayout").data()));
```
If you're able to add a test or two as well, that would be perfect.

**Comment by lovell** [comment]
Please can you target the `yield` branch to ensure your changes are compatible with the forthcoming N-API support.

**Comment by lovell** [comment]
I simplified the tile option parsing in commit 51b1432 on the `yield` branch, so if you rebase against upstream/yield you should hopefully start to see this new option work.

**Comment by lovell** [comment]
v0.25.0 is now available so I've moved this PR back to the master branch.

**Comment by edsilv** [comment]
Hi, nice! Do you need anything further from me? I was a bit out of my depth on the c++ stuff and winded up needing dynamic tiling using [serverless-iiif](https://github.com/nulib/serverless-iiif) (also libvips-based). Looking forward to getting IIIF tiling in Sharp and updating [biiif](https://github.com/edsilv/biiif)!

**Comment by lovell** [comment]
Thanks for the update. Please can you rebase against the latest upstream/master branch and you should the proposed changes in the PR start to work. If you're able to add a unit test or two as well, that would be great, then we can get this merged and out in the next release.

**Comment by edsilv** [comment]
I've pushed the beginnings of a unit test and tried the outputted tiles here: https://repl.it/@edsilv/osd-test

You can see that some tiles are missing when fully zoomed out.

I've also had to add `"preferredFormats": [ "jpeg" ]` to the generated `info.json` as otherwise OSD will look for files ending `.jpg` by default. To get around _that_ you probably would have to do some gymnastics with `IIIFTileSource`, but really it would be better if it just worked out of the box with no customisation. 

Any chance we could change the default outputted tile format to `.jpg` not `.jpeg`?

**Comment by lovell** [comment]
The JPEG suffix logic at https://github.com/lovell/sharp/blob/master/src/pipeline.cc#L936 will need updating to also include `VIPS_FOREIGN_DZ_LAYOUT_IIIF` in the list of layouts for which `.jpg` is used.

**Comment by lovell** [comment]
I had a quick look at this and inverted the JPEG extension logic via commit cf39fc4 to make this a bit simpler, which should hopefully allow you to continue with this PR.

**Comment by edsilv** [comment]
Have tested that locally and the tiles are now .jpg and the generated info.json works without any modifications. However, I'm getting a timeout error when running `npm test`, even if I set the timeout to 5 mins: ` IIIF layout:Error: Timeout of 300000ms exceeded.`
Could this be why I'm seeing missing tiles when fully zoomed out?

**Comment by lovell** [comment]
The "IIIF layout" test is failing as the call to `done()` is commented out (as are the assertions).

https://travis-ci.org/github/lovell/sharp/jobs/662551580#L1596

**Comment by edsilv** [comment]
d'oh - sorry, it was a late night!
That's fixed, but still seeing the bug in OSD: https://youtu.be/69oYSDI22ik

**Comment by lovell** [comment]
Could https://github.com/libvips/libvips/issues/1530 and/or https://github.com/libvips/libvips/issues/1472 relate to this problem?

**Comment by edsilv** [comment]
Some tiles are not available: 
image

**Comment by edsilv** [comment]
(changing `scalefactors` to `scaleFactors` didn't make a difference)

**Comment by edsilv** [comment]
The input image `2569067123_aca715a2ee_o.jpg` is `2725px` in width. The first failing tile in the image above has an `xywh` region of `2048,0,678,2048`. `x:2048 + w:678 = 2726px`, 1px more than the full width of the image. Could this be the problem? Libvips is failing to cut that tile because its bounds exceed the width of the image?

**Comment by lovell** [comment]
Yes, I think this might be a libvips thing. Are you able to reproduce it via the command line?

```sh
vips dzsave input output --layout iiif
```

**Comment by edsilv** [comment]
I downloaded v8.9.1 and am getting the same issue via the command line. I suppose we should flag this with libvips? The explanation above still applies.

**Comment by lovell** [comment]
Yes please, if you could open an upstream issue in libvips that would be great.

It doesn't stop this PR so when you find a moment, please can you uncomment that lovely unit test code, get the assertions passing then we're good to merge, thank you.

**Comment by edsilv** [comment]
Have fixed the tests and left a comment here: https://github.com/libvips/libvips/issues/1472#issuecomment-600270954

File: package.json [hunk 1]
```diff
@@ -135,7 +135,7 @@
   },
   "license": "Apache-2.0",
   "config": {
-    "libvips": "8.9.1"
+    "libvips": "8.9.0"
```
Comment (lovell) [inline package.json]: Please can you remove this change.

File: test/unit/tile.js [hunk 2]
```diff
@@ -91,6 +91,40 @@ const assertGoogleTiles = function (directory, expectedTileSize, expectedLevels,
   });
 };
 
+// const assertIIIFTiles = function (directory, expectedSize, expectedLevels, done) {
```
Comment (lovell) [inline test/unit/tile.js]: Please can you remove this commented-out test code if no longer required.

File: package.json (hunk 1 above)
Comment (edsilv) [reply package.json]: done

File: test/unit/tile.js (hunk 2 above)
Comment (edsilv) [reply test/unit/tile.js]: done

**Comment by lovell** [comment]
Marvellous, thank you Edward, this will be in v0.25.2.

**Pull request merged**